### PR TITLE
Added "Setup" methods to do the Actions input setup that was in BHoM_UI

### PR DESCRIPTION
**Pull request by alelom**
### NOTE: Depends on 

https://github.com/BHoM/BHoM_UI/pull/329
   
### Issues addressed by this PR

Closes https://github.com/BHoM/BHoM_Adapter/issues/259

Moves the logic related to the Adapter Actions setup from the BHoM_UI in the Adapter.  
Advantages:
- Flexibility: Toolkits can now override the new `SetupThen[ActionName]` methods if they need to have different set-up.  
- Consistency: logic related to the adapter is all in one place. See https://github.com/BHoM/BHoM_Adapter/issues/259 for more info.
- Does not require any alignment: all Toolkits will work as before.

I need this quite urgently now because of some needed changes in other toolkits (File_Adapter).

### CODE REVIEW

**Review by adecler — COMMENTED** [review]
Most of the code in here is aimed at handling null cases. So wondering if it wouldn't make more sense to have that directly into to core method instead of a separate one. Genuine question though as I haven't been involved with the adapters long enough that my question might have an obvious answer.

**Comment by alelom** [comment]
> Most of the code in here is aimed at handling null cases. So wondering if it wouldn't make more sense to have that directly into to core method instead of a separate one. Genuine question though as I haven't been involved with the adapters long enough that my question might have an obvious answer.

The main point of having the logic here is to allow overriding by Toolkits. This gives the flexibility of adding checking/verification/additional settings to the "SetUp" methods.

File: BHoM_Adapter/AdapterActions/Pull.cs [hunk 1]
```diff
@@ -42,6 +42,33 @@ public abstract partial class BHoMAdapter
...
+            actualRequest = null;
+
+            if (request is Type)
+                actualRequest = BH.Engine.Data.Create.FilterRequest((Type)request, "");
+
+            if (typeof(IRequest).IsAssignableFrom(request.GetType()))
+                actualRequest = request as IRequest;
+
+            if (request == null)
+                actualRequest = new FilterRequest();
```
Comment (IsakNaslundBh) [inline BHoM_Adapter/AdapterActions/Pull.cs:58]: the order of these two need to switch. Line 54 would crash on null.

Also, would put all if statements in an If else, as only one of them can be true.

File: BHoM_Adapter/AdapterActions/Pull.cs [hunk 2]
```diff
@@ -42,6 +42,33 @@ public abstract partial class BHoMAdapter
...
+
+            return true;
```
Comment (IsakNaslundBh) [inline BHoM_Adapter/AdapterActions/Pull.cs:60]: What happens if I provide a Point? Think this method should raise a warning and return false if non of the above cases is true, and that the method for those cases should return false.

File: BHoM_Adapter/AdapterActions/Move.cs [hunk 3]
```diff
@@ -41,10 +41,22 @@ public abstract partial class BHoMAdapter
...
+        public virtual bool VerifyMoveAdapters(BHoMAdapter source, BHoMAdapter target)
+        {
+            if (source == null || target == null)
+            {
+                Engine.Reflection.Compute.RecordError("Adapter input cannot be null.");
+                return false;
+            }
+
+            return true;
+        }
```
Comment (IsakNaslundBh) [inline BHoM_Adapter/AdapterActions/Move.cs:55]: Do not think we need this method. This is not doing any setup, but simple null checks. To me, this belongs in the Move action.

File: BHoM_Adapter/AdapterActions/Push.cs [hunk 4]
```diff
@@ -35,33 +35,78 @@ namespace BH.Adapter
...
+                objectsToPush = WrapNonBHoMObjects(objects);
+            else
+                objectsToPush = objects.OfType<IBHoMObject>();
+
+            // Clone the objects for immutability in the UI. CloneBeforePush should always be true, except for very specific cases.
+            if (m_AdapterSettings.CloneBeforePush)
+                objectsToPush = objectsToPush.Select(x => x.DeepClone());
+
+            return true;
+        }
```
Comment (IsakNaslundBh) [inline BHoM_Adapter/AdapterActions/Push.cs:85]: As for the adapter verification of the Move, not sure this is a setup. This is _not_ simply converting an object into a request or similar, this is doing stuff I would say is part of the actual push, controlled via action configs and similar. Would not put all this in this separate set up method.

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
Like the action config and request stuff, but think the adapter checking for Move as well as the object checking for push is one step to many. Those to me are clear parts of the push, not something that is simply cleaned up a bit before.

**Comment by IsakNaslundBh** [comment]
> Most of the code in here is aimed at handling null cases. So wondering if it wouldn't make more sense to have that directly into to core method instead of a separate one. Genuine question though as I haven't been involved with the adapters long enough that my question might have an obvious answer.

To expand a bit on alessios answer, it is more about defaults and converts than null checks, and making it possible for each adapter to handle that differently. For example, being able to convert a string, formatted in a specific way into a request that is friendly to that particular adapter, but non-sensical outside that scope.

As for the ActionConfigs, could be used by RevitToolkit for example, that make use of that mechanism a lot, to directly instantiate the appropriate ActionConfig required by a particular action, such as Pull.

File: BHoM_Adapter/AdapterActions/Pull.cs (hunk 2 above)
Comment (IsakNaslundBh) [reply BHoM_Adapter/AdapterActions/Pull.cs:60]: EDIT: If a message is raised elsewhere, or by the UI, simply returning false if it fails is enough I think.

File: BHoM_Adapter/AdapterActions/Push.cs (hunk 4 above)
Comment (alelom) [reply BHoM_Adapter/AdapterActions/Push.cs:85]: Not that I disagree, but this method was already written as a separate method, [`ProcessObjectsForPush()`](https://github.com/BHoM/BHoM_Adapter/blob/master/BHoM_Adapter/AdapterActions/_PushMethods/ProcessObjectsForPush.cs).

It seemed a good opportunity to formalise it better, but then again I understand what you're saying here: the logic that this method contains is strictly related to the Push logic, and therefore it makes more sense as more "embedded" in the Push than separated.

I will restore it as it was before.

File: BHoM_Adapter/AdapterActions/Pull.cs [hunk 5]
```diff
@@ -42,6 +42,33 @@ public abstract partial class BHoMAdapter
...
+        public virtual bool SetupPullRequest(object request, out IRequest actualRequest)
+        {
+            // Always assure there is a Request. Allow to input a Type to generate a FilterRequest.
+            actualRequest = null;
+
+            if (request is Type)
+                actualRequest = BH.Engine.Data.Create.FilterRequest((Type)request, "");
+            else if (request == null)
+                actualRequest = new FilterRequest();
+            else if (typeof(IRequest).IsAssignableFrom(request.GetType()))
```
Comment (IsakNaslundBh) [inline BHoM_Adapter/AdapterActions/Pull.cs:55]: Think the line you had worked, but for consistency with the type checking above, would change to this

```suggestion
            else if (request is IRequest)
```

File: BHoM_Adapter/AdapterActions/Pull.cs [hunk 6]
```diff
@@ -42,6 +42,33 @@ public abstract partial class BHoMAdapter
         /* These methods represent Actions that the Adapter can complete. 
            They are publicly available in the UI as individual components, e.g. in Grasshopper, under BHoM/Adapters tab. */
 
+        [Description("Performs a set up for the Request input of the Pull Action.")]
+        public virtual bool SetupPullRequest(object request, out IRequest actualRequest)
+        {
+            // Always assure there is a Request. Allow to input a Type to generate a FilterRequest.
+            actualRequest = null;
+
+            if (request is Type)
```
Comment (IsakNaslundBh) [inline BHoM_Adapter/AdapterActions/Pull.cs:51]: this is now jsut details, but as I commented below anyway, would change the order of these again, mostly just for clarity to:

1. check null
2. check if request
3. check if type

Mostly as an order of "importance" and for general validation purpose. as in, first check for null, then check if it is the exact thing you are after, then check for any additional possibility (in this case the type)

File: File_Adapter/AdapterActions/Push.cs [hunk 7]
```diff
@@ -60,13 +57,9 @@ public override List<object> Push(IEnumerable<object> objects, string tag = "",
...
 
-            if (objectsToPush.Count() != objects.Count())
-                Engine.Reflection.Compute.RecordWarning("The file adapter can currently only be used with BHoMObjects." + Environment.NewLine +
-                    "If you want to push non-BHoMobject, specify a push config with the option `WrapNonBHoMObject` set to true.");
-
-            bool success = this.FullCRUD(objectsToPush, pushType, tag, actionConfig);
+            bool success = this.FullCRUD(objects.OfType<IBHoMObject>(), pushType, tag, actionConfig);
 
-            return success ? objectsToPush.Cast<object>().ToList() : new List<IObject>().Cast<object>().ToList();
+            return success ? objects.ToList() : new List<IObject>().Cast<object>().ToList();
```
Comment (IsakNaslundBh) [inline File_Adapter/AdapterActions/Push.cs:62]: not something changed here, but cant it just be:

```suggestion
            return success ? objects.ToList() : new List<object>();
```

not sure I get the casting and to list thing there? Might be leftover from previous iterations?

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
some small additional tweaks

File: BHoM_Adapter/AdapterActions/Pull.cs (hunk 6 above)
Comment (alelom) [reply BHoM_Adapter/AdapterActions/Pull.cs:51]: Sounds good, done.

File: File_Adapter/AdapterActions/Push.cs [hunk 8]
```diff
@@ -37,10 +37,7 @@ public override List<object> Push(IEnumerable<object> objects, string tag = "",
         {
             // --------------- SET-UP ------------------
 
-            // Process the objects (verify they are valid; DeepClone them, wrap them, etc).
-            IEnumerable<IBHoMObject> objectsToPush = ProcessObjectsForPush(objects, actionConfig); // Note: default Push only supports IBHoMObjects.
```
Comment (IsakNaslundBh) [inline File_Adapter/AdapterActions/Push.cs:41]: Should this part really be deleted now?

File: File_Adapter/AdapterActions/Push.cs [hunk 9]
```diff
@@ -60,14 +57,9 @@ public override List<object> Push(IEnumerable<object> objects, string tag = "",
 
             CreateFileAndFolder(pushType);
 
-            if (objectsToPush.Count() != objects.Count())
```
Comment (IsakNaslundBh) [inline File_Adapter/AdapterActions/Push.cs:63]: same here, should this really be removed now that we are keeping the pre-processing as before?

**Review by IsakNaslundBh — APPROVED** [review]
now LGTM. Think this is a good balance!

**Pull request merged**